src/attpc_engine/detector/simulator.py

```diff
@@ -0,0 +1,478 @@
+            [
+                normalvariate(point, np.sqrt(params.detector.fano_factor * point))
+                for point in electrons
+            ]
+        )
+
+        # Must have an integer amount of electrons at each point
+        fano_adjusted = fano_adjusted.astype(np.int64)
+
+        # Remove points in trajectory that create less than 1 electron
+        mask = fano_adjusted >= 1
+        self.track = self.track[:, mask]
+        fano_adjusted = fano_adjusted[mask]
+
+        return fano_adjusted
+
+    def z_to_tb(self, params: Parameters):
+        """
+        Converts the z-coordinate of each point in the track
+        from physical space (m) to time (time buckets). Note that
+        the time buckets returned are floats and will be rounded
+        down to their correct bins when the electrons at each
+        point are transported to the pad plane.
+
+        Parameters
+        ----------
+        params: Parameters
+            All parameters for simulation.
+
+        Returns
+        -------
+        np.ndarray[floats]
+            1xN array of time buckets for N points in the track
+            that produce one or more electrons.
+        """
+        # Z coordinates of points in track
+        zpos: np.ndarray = self.track[2]
+
+        dv: float = params.calculate_drift_velocity()
+
+        # Convert from m to time buckets
+        zpos = np.abs(zpos - params.detector.length)  # z relative to the micromegas
+        zpos /= dv
+        zpos += params.electronics.micromegas_edge
+
+        return zpos
+
+    def generate_hits(self, params: Parameters):
+        """
+        Finds the pads hit by the electrons transported from each point
+        of the nucleus' trajectory to the pad plane.
+
+        Parameters
+        ----------
+        params: Parameters
+            All parameters for simulation.
+
+        Returns
+        -------
+        dict[int: np.ndarray[int]]
+            Dictionary of pads hit by transporting the liberated electrons.
+            The key is the pad number and the value is a 1xNUM_TB array
+            of the number of electrons detected at each time bucket.
+        """
+        results = {}
+
+        # Apply gain factor from micropattern gas detectors
+        electrons: np.ndarray = self.electrons * params.detector.mpgd_gain
+
+        # Find time bucket of each point in track
+        point_tb: np.ndarray = self.z_to_tb(params)
+
+        for idx, tb in enumerate(point_tb):
+            point = TransportPoint(
+                params.pad_map,
+                params.pads.map_params,
+                params.detector.diffusion,
+                params.detector.efield,
+                params.calculate_drift_velocity(),
+                (self.track[0][idx], self.track[1][idx]),
+                electrons[idx],
+                tb,
+            )
+            intermediate: dict[int : np.ndarray] = point.pads_hit
+            results = merge_dicts(results, intermediate)
```

Comment:
I think we could probably simplify this, but I need to test some stuff to make sure!